### ISSUE TITLE
Proposal: Identify monitored apps by bound service

### ISSUE THREAD

**Issue by eloo** [promregator/promregator#61]
Hi,

today i took some time to think about how the identification of monitored apps could be improved.
The problem with regex or all i see is that it is very static and must maybe be adapted from time to time from someone who has access and the know how for the infrastructure/kubernetes.

So my first tought was about if it would be possible to "mark" apps to be monitored, e.g. with a label. Then promregator can discover all apps with that label. So far so good, but labelling is not possible as far as i know.

So the idea of user-provided-services came up in my mind and i guess this could be a really cool solution for this problem. Also it must not be exactly "user-provided".

So my idea is about to let promregator monitor all apps which has bound a predefined service (which can be a UPS or created by a service-broker).
So for example the app1 has bound the UPS monitoring. Then promregator is fetching all apps which has bound the monitoring service and is afterwards scraping the metrics from these apps.
And also to solve the problem with specific scrape configs for different apps (e.g. different basic auth creds) the apps could may expose the necessary information as environment variables (which i hope can be fetched by cf-api?).

I know this will cause some major changes but for the future and especially for greater setup i guess this could really improve the usability of promregator.

So @eaglerainbow what do you think? And is this technically possible at all?

Thanks

**Comment by eaglerainbow** [comment]
I think I can understand your motivation quite well. In fact...

> So far so good, but labelling [in CF] is not possible as far as i know.

... had been my preferred choice, when I initially started to develop promregator. However, I also could not find a way to tag applications explicitly... and even worse, there was not appropriate query mechanism provided in the API.  I then started to google on that, too, and there several pages suggest to "simulate tagging" by...

> So my idea is about to let promregator monitor all apps which has bound a [user] predefined service

... which I had a brief look into as well.
There problem is that you can't query them via the CF API either. It would mean that you have to iterate over all orgs and all spaces, read their metadata and query them for certain UPSes to exist or not. If one exists, you then would have to reverse-travel to the applications (bindings). Afterwards, you have to play the same game as we currently do: You have to analyze the domain association, etc...

In general, Promregator can do everything on the CF API, which the CF API permits, the underlying `cf-java` library supports and the user is authorized for. However, we have some very strict constraint - and that is performance: User usually want to have scraping intervals around 10-15 seconds. And strangely, they expect that for each scraping interval, they get a "consistent set of data" from the platform. 
Promregator's main challenge already today is not the scraping itself (that is comparably cheap), but it's the discovery (that's why we already invested a huge bunch of effort into caching)! Labelling via services would make things even more complicated, as the number of API requests to send to the CF API would even increase significantly - not to mention that you would have to do this kind of scanning very often, as new applications could be popping up and should be subject to scraping...

> I know this will cause some major changes but for the future and especially for greater setup i guess this could really improve the usability of promregator.

With that I totally agree. But I did not find a good way to do it...
[BTW: There is more to this: who tells you that the scraper and the application developer is the same party? Who tells you that the application **wants** to bind to such a service - or even can be reached to be asked?)

> So, eaglerainbow, what do you think? 

I like the idea, but I have not found the silver-bullet for the design.

> And is this technically possible at all?

Maybe. Major challenge is caching/performance during discovery (as laid out above).

However, I would like to encourage you to follow up on this idea (as I really see the great potential in it). That is why I would be more than happen, if someone would prototype/build a PoC of that and report on the findings.

**Comment by eaglerainbow** [comment]
## Call for ideas
I'd like to announce a "Call for Ideas" on this topic here in this thread. So far, we haven't found the right approach of doing it, but the intention appears to me worth a deep concept/design discussion. 

So, **to everybody**, in case you have an idea how we would make it, please post your idea (and may it still appear to be a wild one...)!

**Comment by poblin-orange** [comment]
Nice ideas.

Some points:
- an OSB Open Service Broker is a great idea. This standard is also now also available on K8s (catalog).
- there is good support for java OSB broker implementation https://github.com/spring-cloud/spring-cloud-open-service-broker
- a non privileged Cloud Foundry user can register an OSB broker localy to his space
- the service binding could trigger some buildpack automation mechanism (ie: java buildpack has Metric Writer Metrics, with Spring boot support to prometheus)
- wrt implementation and persistence, we are working on a generic OSB solution, able to deploy to Bosh / and or k8s. This might be usefull to manage persistence AND a complete dedicated prometheus ondemand for the OSB broker (ie: bosh deploy on the fly the prometheus bosh release, wired to promgregator, available though the service instance dashboard).   https://github.com/orange-cloudfoundry/cf-ops-automation-broker

**Comment by eaglerainbow** [comment]
Thanks for your ideas. 

> an OSB Open Service Broker is a great idea. This standard is also now also available on K8s (catalog)

As Promregator is a specific for Cloud Foundry applications anyway, having an [OSB](https://github.com/openservicebrokerapi/servicebroker/) is a nice thing, but it won't be required on K8s anyway (for k8s there are alternative approaches available in Prometheus). Therefore, a "simple-and-stupid CF SB" might already do the trick.

> a non privileged Cloud Foundry user can register an OSB broker localy to his space

The question here is what the OSB needs to contain. If it requires a running database in the background, the TCO for running a SB instance for each space might be technically possible, but not viable, as the costs of operations will be too high.

(BTW: does anyone know why there is only a `--space-scoped` and not also a `--org-scoped` in CF?)

> the service binding could trigger some buildpack automation mechanism

Yep, that might be helpful to - for example - trigger the configuration of exposure in Micrometer (for instance). However, as this happens on the target application's side, it does not help Promregator...

> This might be usefull to manage persistence AND a complete dedicated prometheus ondemand for the OSB broker

Mmmh... I doubt that you want to have an own instance of Prometheus running for each space-scoped instance of a "PSB". I would expect that administrators would prefer to have Prometheus stationed more central for their deployment.  --- Or did I get your argument somehow wrong?

**Comment by poblin-orange** [comment]
The TCO for database prometheus is a recurrent point. However, having logically dedicated bakend services is a requirement for operators beyond POC .

Can think of 2 uses cases 
-  public CF (ie PWS). In that case, operators get an org. A their only way to extend is using the space scoped broker, and can deploy their own OSB as a cf app. The OSB app can it self leverage sql service (say p-mysql on PWS) for its persistence. But no way to host Prometheus, as its not in the cf provider catalog.

- on premise use case. Here, we have plateform operators role, and we can extend the market place with a full scale OSB. Moreover, we can use any solution for backends (prometheus / sql db), including low footprint solutions like k8S to host them. This is what we're trying to build with COAB. cf apps operators simply use OSB service, and get a dedicated metrics solution. 

Globally, we also want to minimize code level instrumentation, this is why i referred to buildpacks. On the shelf java buildpack already has built in support for micro-meters. This support can be activated with properties set by OSB binding properties

https://github.com/cloudfoundry/java-buildpack/releases/tag/v4.8
https://github.com/cloudfoundry/java-buildpack-metric-writer/blob/master/java-buildpack-metric-writer-boot-2/src/test/java/org/cloudfoundry/metrics/boot2/MicrometerMetricWriterTest.java

**Comment by bkez322** [comment]
I would be very interested in such a feature.

I am an administrator of Cloud Foundry, and want to make it as easy as possible for my app-developers to collect app metrics with Prometheus. My thoughts are particularly focused on that use-case. These thoughts have largely already been mentioned in this thread, but I'm just summarizing how I'd expect the feature to work. :)

I would like there to be extremely simple and *stateless* Prometheus service broker. This service broker shouldn't do anything, other than provide `200` status codes for all the Service-Broker API paths. I don't think it needs to store anything, because I believe CF itself already stores records of all service instances and service bindings.

I would imagine that I, as an administrator, would have a single Prometheus (or multiple identical ones) stood up ahead of time, and this single Prometheus would service all apps on the foundation. I would not expect the service broker to be provisioning new Prometheus's.

One significant thing to mention is that users should have the ability to specify the endpoint their app is exposing metrics on. SpringBoot1.x defaults exposes metrics `/prometheus`, SB2.x on `/actuator/prometheus`, and other apps on `/metrics`. Perhaps users would set this as an environment variable.

I personally do not particularly care about letting users specify service plans, though that would be a nice-to-have if someday there were to be a more elaborate service broker that provisions new Prometheus's.

There would also probably be an environment variable for the password Prometheus should use to scrape a particular app.

So after a user binds their app, Promregator would of course only scrape from the apps bound to a particular service, and using the app-specific endpoints and passwords specified (and I realize this is harder than it sounds from a efficiency standpoint).

**Comment by eaglerainbow** [comment]
Thanks for your input!

> These thoughts have largely already been mentioned in this thread, but I'm just summarizing how I'd expect the feature to work. :)

Sounds familiar with all the other posts here: We all can describe the idea quite well, but the way to heaven is unclear 😀 

> This service broker shouldn't do anything, other than provide 200 status codes for all the Service-Broker API paths. I don't think it needs to store anything, because I believe CF itself already stores records of all service instances and service bindings.

Mhmm. What's then the benefit of the service broker compared to the user-provided-service eloo has suggested initially? If the PSB (Promregator Service Broker) does not store anything, it also cannot deal as "one-stop-shop location" for fetching all active targets. Instead, we would have to ask the Cloud Controller "somehow" about all service instances - to retrieve the scraping metadata (e.g. name of the endpoint, username, password, ...) from there again. In short, we would have won little with the PSB then...
Or did I miss something?

**Comment by eaglerainbow** [comment]
Reviewing all the ideas again, I just stumbled over https://apidocs.cloudfoundry.org/1.23.0/user_provided_service_instances/list_all_user_provided_service_instances.html in the CF API. With that one you may display all UPS across all spaces and all orgs. With https://apidocs.cloudfoundry.org/1.23.0/user_provided_service_instances/list_all_service_bindings_for_the_user_provided_service_instance.html it's also possible to retrieve all applications bound to these service instances. This could allow the following approach:

1. Users have to create a UPS where the credentials JSON has to be of a certain format. That format must uniquely identify that the JSON is intended to be used by Promregator. The JSON file may contain attributes like 
   * the endpoint's name
   * username
   * password
2. Users bind one or more applications to the UPS. A binding indicates that the user expect these applications to be scraped.
3. Promregator queries for **all** UPSes on the platform (also those, which it is not interested in). 
4. Scanning all the credentials JSON documents, it filters out those, which do not comply to the Promregator's format. The remaining rest is accepted and parsed accordingly.
5. For each "promregator-valid" UPS, the service bindings are retrieved. Promregator gets to know the application Id which needs to be scraped. (BTW: the UPS also tells Promregator in which space the application is in).
6. Promregator has to start discovery now "bottom-up": The name of the space and the org has to be determined and "spaceSummary" needs to be triggered to find the number of instances and the routes/hostnames to be used for scraping. 

That's the first time I see that this UPS approach could work out.
However, the argument

> Labelling via services would make things even more complicated, as the number of API requests to send to the CF API would even increase significantly ➡️ performance

still holds true with it. Yet, I see a small chance for that:
* The first request to fire would be "list all UPS". The response already contains the credentials JSON document, so we could determine already at this stage, which service instances are ok.

Immediately, we could start three streams in parallel: 
1. The "list of UPS" also provides the space ID. With that we can query the space name and the associated org Id. Again with the orgId we could query for the name of the org.
2. With the list of promregator-valid UPSes, we can query their service instances. We can do so in one big request (multiple-select-single) so that we determine the applicationIds.
3. Based on the list of space Ids, we are able to immediately fire the spaceSummary calls, which are most likely the most costly one's. 

Once all these three streams have finished, we are able to fiddle all the stuff together. 

As we could do all this in parallel, it seems reasonable to me to give it a try at least.

What will be a challenge for sure, is error handling in all these three streams: 
* Stream 1 is not error prone (otherwise we have an inconsistency on the platform anyway). 
* Stream 2 must be able to handle the case that no service binding has been created. It also needs to survive the case that someone has deleted the UPS, which we just retrieved, but the request of the service binding then will fail due to the deletion of the UPS.
* Stream 3 can do best effort.
* Bringing all the data together, however, will be tricky: 
    * You may have service bindings to applications, which do not appear in the spaceSummary, 
    * You may have UPS, for which you cannot retrieve a spaceSummary
    * ...


Is anyone willing to implement/support me with a proof of concept?

**Comment by bkez322** [comment]
After a little more thought, using UPS is probably the better route over a service broker. UPS is much simpler.

The nice thing about the service broker is you can advertise prometheus in the marketplace, but that alone isn’t nearly a good enough reason to justify its existence, I think.

**Comment by eaglerainbow** [comment]
... and moreover, having an implementation for UPS does not mean that we never can go the service broker approach later on.

I am currently experimenting with the UPS approach. Unfortunately, `cf-java-client` does not have some mass functions available (e.g. no mass-retrieval of UPS bindings), such that we fall back to single request modes again. This could cost a lot of performance. Perhaps we have to request a feature at them first...

**Comment by eaglerainbow** [comment]
I have uploaded the current state of my experiments at branch [upsDiscovery](https://github.com/promregator/promregator/tree/upsDiscovery). Always latest diff can be generated [here](https://github.com/promregator/promregator/compare/upsDiscovery?expand=1).
Short summary of current state:
* Everything is on single-request mode for UPS discovery.
* No error handling implemented yet.
* `applicationIdFilter` and `instanceFilter` are not considered yet (and pose a problem as far as I can see).
* Basic Authentication may work already.
* `Instance` moved to `scanner` package and heavily refactored.
* Several implementations of `CFAccessor` just got dummy implementations to make things compile again.
* No unit test coverage at all.

I think, this test coding needs to take a little rest: The approach invalidates many things we have done so far. So having a second look with a little distance is paramount.

**Comment by eaglerainbow** [comment]
I am positively surprised: I did some performance measurements on first initial request (cold cache) based on the current state of the `CFAccessorSimulator` (2d2478ca93e23f6cb9fdb7b381168f6ee2e7717b) using the new "ups-based approach" comparing it with the "classical way" we have:

classical approach:
* 100 applications in one space
* 2 instances for each application
* the space is in one org

UPS-based approach:
* 100 applications attached to two UPSes like this:
* one UPS with promregator data local to the application (e.g. we have 100 service instances for this)
* one "global UPS" with promregator data, one binding from each application to that global UPS.
* 2 instances for each application
* all applications in the same space
* the space is in one org.

The latency values are interesting:
* "classical way": 1200 - 1500 ms
* UPS-based approach: 580 - 610 ms

Note that the simulator creates a random load delay of 0 - 250ms (avg. 125ms) per roundtrip to the "Cloud Controller".
Note that the calls for the UPS-based approach are not cached yet.
I was also able to make the two approaches run in parallel side-by-side. Except for the `spaceSummary` calls they are distinct.

**Comment by eaglerainbow** [comment]
The decreasing performance when introducing caching mainly seems to have come from bad luck with the random load delay.

**Comment by eaglerainbow** [comment]
Small performance test of the platform:

Data was generated with the following script:
```bash
for i in `seq 1 100`; do cf cups promregator-$i -p '{"promregator-version":"1","path":"/testmetrics","username":"user","password":"pw"}'; done
```

Test approach `time curl --header "Authorization: bearer ey..."`

Datapoints:
```
real	0m2.658s
user	0m0.024s
sys	0m0.024s

real	0m1.761s
user	0m0.020s
sys	0m0.008s

real	0m2.060s
user	0m0.036s
sys	0m0.008s

real	0m1.319s
user	0m0.032s
sys	0m0.016s

real	0m1.756s
user	0m0.008s
sys	0m0.008s
```
BTW: Data volume received: ~51kb

Enhancement of test case:
```
for i in `seq 2 100`; do cf delete-service -f promregator-$i; done
```
=> only one CUPS instance anymore

Datapoints:
```
real	0m0.798s
user	0m0.036s
sys	0m0.008s

real	0m1.017s
user	0m0.020s
sys	0m0.016s

real	0m2.302s
user	0m0.032s
sys	0m0.020s

real	0m6.389s
user	0m0.036s
sys	0m0.008s

real	0m0.933s
user	0m0.012s
sys	0m0.008s

real	0m0.845s
user	0m0.012s
sys	0m0.008s

real	0m1.396s
user	0m0.016s
sys	0m0.004s
```
The 6s+ data point looks to be an out-lier. 

Bottomline: 100 CUPS cost ~1s of response time.

**Comment by eaglerainbow** [comment]
Mmhmhm.. automatic cleanup fails with
```
org.springframework.jms.support.converter.MessageConversionException: Could not map JSON object [org.cloudfoundry.promregator.discovery.UPSBasedInstance@64840084]; nested exception is com.fasterxml.jackson.databind.exc.InvalidDefinitionException: No serializer found for class org.cloudfoundry.promregator.auth.BasicAuthenticationEnricher and no properties discovered to create BeanSerializer (to avoid exception, disable SerializationFeature.FAIL_ON_EMPTY_BEANS) (through reference chain: org.cloudfoundry.promregator.discovery.UPSBasedInstance["authenticationEnricher"])
```
=> There is still more work to be done...

**Comment by eloo** [comment]
@eaglerainbow really cool that you are working on this idea.. 
your approach sounds really good..
but i'm getting a little headache when i read this 
> Promregator queries for all UPSes on the platform (also those, which it is not interested in).

do you know if this could be limited to only spaces i've access to? or is it already the case? because out of the security view this will be a showstoper for us
maybe you can test this?

thx

**Comment by eaglerainbow** [comment]
>> Promregator queries for all UPSes on the platform (also those, which it is not interested in).

> but i'm getting a little headache when i read this

Me, too. But less of security (reasoning see below), but more about performance: Consider that we have thousands of UPSes on the platform, but only a dozen (or so) really are relevant to Promregator. Can the CF Cloud Controller really answer the metadata requests fast enough? (we thought similar things about the spaceSummary request before - you know what kind of trouble we have there...)

> do you know if this could be limited to only spaces i've access to? or is it already the case? 

The CF Cloud Controller always answers all requests only for those spaces for which you have been granted access. All other must be filtered anyway. Anything else would be less of a problem for Promregator, but would be a major security issue for the platform, as any REST-based client then could gain information from a foreign domain. 
Promregator in no case has any additional privilege on the platform than those privileges which the user has been granted on platform. Or in other words: Promregator has no elevated rights on the platform, but *always* like any other REST-based client would do (e.g. the CF CLI).

> because out of the security view this will be a showstoper for us 

You are safe on this. If you have any hint that this might be violated, we'll immediately create a security incident on the Cloud Controller component.

> maybe you can test this?

I tested it with my own account, which runs on Bluemix. I could only see what I am allowed to see plus it is the very same thing that you will also see when using the CF CLI. 

If we are already talking about security on this feature proposal... there is one more thing with security, which is a little awkward with this approach, but I have no clue how to prevent it: We are requesting all UPSes on all spaces in all orgs (for which logged on user has access to). This means that these are both those UPSes, which are intended to be used by Promregator, but also those which Promregator isn't interested at all. Unfortunately, with the initial request, also the entire content (that is to say: the credentials of **all** UPSes) is transferred over the wire. In other words: Promregator also gets sent all the credentials of all UPSes - also those, which Promregator isn't interested in. So far, I did not find any way to prevent that.
The current mitigation approach is that we try to dispose those credentials, which we are **not** interested in, as soon as possible (early prefiltering). That has also the very nasty side-effect that we virtually cannot request-cache the native UPS service call and the UPS binding call. 😢 
However, Promregator could still be subject to a attack vector when trying to sniff the credentials (Notabene: I don't have any indication that the approach would be *vulnerable* to such an attack - I am just speaking about on a conceptual security topic here - it's always better to *not* have some critical data than having it 😁 ).

**Comment by eaglerainbow** [comment]
For the record: The issue with 
```
org.springframework.jms.support.converter.MessageConversionException: Could not map JSON object [org.cloudfoundry.promregator.discovery.UPSBasedInstance@64840084]; nested exception is com.fasterxml.jackson.databind.exc.InvalidDefinitionException: No serializer found for class org.cloudfoundry.promregator.auth.BasicAuthenticationEnricher and no properties discovered to create BeanSerializer (to avoid exception, disable SerializationFeature.FAIL_ON_EMPTY_BEANS) (through reference chain: org.cloudfoundry.promregator.discovery.UPSBasedInstance["authenticationEnricher"])
```
is resolved with 16983532974d77aa46e4fb33e2383653071b9d20 by changing the type of the `DISCOVERER_INSTANCE_REMOVED` JMS message.
It's a hard way to learn our living in this case here...

**Comment by eaglerainbow** [comment]
Another aspect to consider on this: The numbers of UPS may grow huge. They may not fit on a single page anymore. However, CF API is page-based...

**Comment by eaglerainbow** [comment]
I just found another issue of this issue by chance:
* Today, the user, which is used for fetching the application data, can be limited to the role "Space Auditor". 
* With the UPS-based approach in place, the user must be able to look into the UPSes content.
* However, for that at least the role "Space Developer" is needed.
* "Space Developers" in turn, though, are able to stop applications, restage them, push them, delete them (or even services, which could be databases, ....)

W.r.t. security, this finding causes yet-another-challenge to this approach.
